File: 53.maximum-subarray.py
#
# @lc app=leetcode id=53 lang=python3
#
# [53] Maximum Subarray
#

# @lc code=start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        mxsub = cursum = nums[0]

        for num in nums[1:]:
            cursum = (cursum + num) if cursum > 0 else num
            mxsub = max(mxsub, cursum)
        return mxsub


# @lc code=end

File: RustSolutions/src/53.maximum-subarray.rs
/*
 * @lc app=leetcode id=53 lang=rust
 *
 * [53] Maximum Subarray
 */

// @lc code=start
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        nums.iter().fold((0, i32::min_value()), |(sum, mxsub), &num| {
            let sum = if sum > 0 { sum + num } else { num };
            (sum, mxsub.max(sum))
        }).1
    }
}
// @lc code=end

